### ISSUE TITLE
configure cross-compiling fails in new versions

### ISSUE THREAD

**Issue by parduz** [wxWidgets/wxWidgets#22886]
### Description
Using this command:
```../configure --disable-shared --disable-debug --host=arm-linux-gnueabihf```
**_configure_** fails in version 3.2.1 and 3.1.7, while it succeeds with version 3.1.4 and older.

**_configure_** fails in testing GTK:

>   checking for GTK+ version...
>   checking for arm-linux-gnueabihf-pkg-config... (cached) /usr/bin/arm-linux-gnueabihf-pkg-config
>   checking pkg-config is at least version 0.16... yes
>   checking for GTK+ - version >= 3.0.0... Package gtk+-3.0 was not found in the pkg-config search path.
>   Perhaps you should add the directory containing `gtk+-3.0.pc'
>   to the PKG_CONFIG_PATH environment variable
>   No package 'gtk+-3.0' found
>   Package gthread-2.0 was not found in the pkg-config search path.
>   Perhaps you should add the directory containing `gthread-2.0.pc'
>   to the PKG_CONFIG_PATH environment variable

and after other errors:

> *** Could not run GTK+ test program, checking why...
*** The test program failed to compile or link. See the file config.log for the
*** exact error that occurred. This usually means GTK+ is incorrectly installed.
checking for gtk-config... no
checking for GTK - version >= 1.2.7... no
*** The gtk-config script installed by GTK could not be found
*** If GTK was installed in PREFIX, make sure PREFIX/bin is in
*** your path, or set the GTK_CONFIG environment variable to the
*** full path to gtk-config.
checking for gtk-config... (cached) no
checking for GTK - version >= 1.2.3... no
*** The gtk-config script installed by GTK could not be found
*** If GTK was installed in PREFIX, make sure PREFIX/bin is in
*** your path, or set the GTK_CONFIG environment variable to the
*** full path to gtk-config.
configure: error:
The development files for GTK+ were not found. For GTK+ 2, please
ensure that pkg-config is in the path and that gtk+-2.0.pc is
installed. For GTK+ 1.2 please check that gtk-config is in the path,
and that the version is 1.2.3 or above. Also check that the
libraries returned by 'pkg-config gtk+-2.0 --libs' or 'gtk-config
--libs' are in the LD_LIBRARY_PATH or equivalent.

The same command works flawlessy with wx version 3.1.4 and 3.1.2; i then can build the widgets and the _minimal_ sample app.

### Platform and version information

- wxWidgets version 3.2.1 and 3.1.7, source files for linux.
- tested and verified on two PCs: one Debian11 in a Windows 10 WSL 2, one Debian 9 32bit + Gnome
- Compiler being used: gcc

**Comment by vadz** [comment]
What does `/usr/bin/arm-linux-gnueabihf-pkg-config --cxxflags gtk+-3.0` return for you? Could you please look at the relevant parts of `config.log` and/or compare the working `config.log` from 3.1.4 with the non-working one from the later version?

I suspect this might be due to the changes of 893ebbab0c (Disable use of build system pkg-config files when cross-compiling, 2020-12-13) and that the solution is probably to extend the test for `PKG_CONFIG_LIBDIR` to also check for arch-spefific pkg-config being used.

**Comment by parduz** [comment]
> What does /usr/bin/arm-linux-gnueabihf-pkg-config --cxxflags gtk+-3.0 return for you?

`$ /usr/bin/arm-linux-gnueabihf-pkg-config --cflags gtk+-3.0

-pthread -I/usr/include/gtk-3.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/arm-linux-gnueabihf/dbus-1.0/include -I/usr/include/gtk-3.0 -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/glib-2.0 -I/usr/lib/arm-linux-gnueabihf/glib-2.0/include`



> Could you please look at the relevant parts of config.log and/or compare the working config.log from 3.1.4 with the non-working one from the later version?

I hope this is what you asked for:

**v.3.1.4**
`GTKPRINT_CFLAGS='-pthread -I/usr/include/gtk-3.0/unix-print -I/usr/include/gtk-3.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/arm-linux-gnueabihf/dbus-1.0/include -I/usr/include/gtk-3.0 -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/glib-2.0 -I/usr/lib/arm-linux-gnueabihf/glib-2.0/include'
GTKPRINT_LIBS='-L/usr/lib/arm-linux-gnueabihf -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0'
GTK_CFLAGS='-pthread -I/usr/include/gtk-3.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/arm-linux-gnueabihf/dbus-1.0/include -I/usr/include/gtk-3.0 -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/glib-2.0 -I/usr/lib/arm-linux-gnueabihf/glib-2.0/include'
GTK_CONFIG=''
GTK_LIBS='-L/usr/lib/arm-linux-gnueabihf -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lgthread-2.0 -pthread -lglib-2.0 -lX11'
`
**v3.1.7**
`GTKPRINT_CFLAGS=''
GTKPRINT_LIBS=''
GTK_CFLAGS=''
GTK_CONFIG='no'
GTK_LIBS=''
`


> I suspect this might be due to the changes of ...

There's a temp solution, or even a hack, i could use to be able to bypass the problem?

**Comment by vadz** [comment]
So it looks like using `/usr/bin/arm-linux-gnueabihf-pkg-config` should work, thanks. Where does this file come from BTW? I don't see it in (non-ARM) Debian repositories. I'd like to install it here to be able to test myself.

> I hope this is what you asked for:

Not quite, I wanted to see what is configure doing when trying to find this package, i.e. which `pkg-config`, e.g. when configuring normally here I have something like this in my `config.log`:

```
configure:26893: checking for GTK+ version
configure:26902: result:
configure:26995: checking for pkg-config
configure:27025: result: /usr/bin/pkg-config
configure:27050: checking pkg-config is at least version 0.16
configure:27053: result: yes
configure:27067: checking for GTK+ - version >= 3.0.0
configure:27174: gcc -o conftest -I/usr/include/x86_64-linux-gnu -D_GNU_SOURCE  -pthread -I/usr/include/gtk-3.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/gtk-3.0 -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/x86_64-linux-gnu -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include   conftest.c -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lgthread-2.0 -pthread -lglib-2.0 -lcurl  -lm >&5
configure:27174: $? = 0
configure:27174: ./conftest
configure:27174: $? = 0
configure:27188: result: yes (version 3.24.34)
```

(BTW, please notice that you need to use triple backticks around verbatim blocks of text in GitHub Markdown, I've taken the liberty to edit your reply to add them as otherwise it wasn't very readable).

> There's a temp solution, or even a hack, i could use to be able to bypass the problem?

You could try reverting the commit mentioned above -- does it work for you?

**Comment by parduz** [comment]
> So it looks like using `/usr/bin/arm-linux-gnueabihf-pkg-config` should work, thanks. Where does this file come from BTW? I don't see it in (non-ARM) Debian repositories. I'd like to install it here to be able to test myself.

these should be the relevant commands from _history_:
```
sudo apt-get install libgtk-3-dev build-essential checkinstall
sudo apt install gcc make gcc-arm-linux-gnueabihf binutils-arm-linux-gnueabihf
sudo dpkg --add-architecture armhf
sudo apt-get update
sudo apt install libgtk-3-dev:armhf
```


> Not quite,

Sorry. Let me try again:
**v3.1.4**
```
configure:21330: checking for arm-linux-gnueabihf-pkg-config
configure:21348: found /usr/bin/arm-linux-gnueabihf-pkg-config
configure:21360: result: /usr/bin/arm-linux-gnueabihf-pkg-config
configure:21428: checking pkg-config is at least version 0.9.0
configure:21431: result: yes

...

configure:22875: checking for GTK+ version
configure:22884: result: 
configure:22935: checking for arm-linux-gnueabihf-pkg-config
configure:22965: result: /usr/bin/arm-linux-gnueabihf-pkg-config
configure:23033: checking pkg-config is at least version 0.16
configure:23036: result: yes
configure:23050: checking for GTK+ - version >= 3.0.0
configure:23171: result: yes (version 3.24.24)
configure:26298: checking for X11/Xlib.h
configure:26298: arm-linux-gnueabihf-gcc -c -D_GNU_SOURCE   conftest.c >&5
configure:26298: $? = 0
configure:26298: result: yes
configure:26311: checking for X11/XKBlib.h
configure:26311: arm-linux-gnueabihf-gcc -c -D_GNU_SOURCE   conftest.c >&5
configure:26311: $? = 0
configure:26311: result: yes
configure:26715: checking for arm-linux-gnueabihf-pkg-config
configure:26745: result: /usr/bin/arm-linux-gnueabihf-pkg-config
configure:26813: checking pkg-config is at least version 0.9.0
configure:26816: result: yes

```

**v3.1.7**
```
configure:25712: checking for arm-linux-gnueabihf-pkg-config
configure:25730: found /usr/bin/arm-linux-gnueabihf-pkg-config
configure:25742: result: /usr/bin/arm-linux-gnueabihf-pkg-config
configure:25810: checking pkg-config is at least version 0.9.0
configure:25813: result: yes

...

configure:27572: checking for GTK+ version
configure:27581: result: 
configure:27633: checking for arm-linux-gnueabihf-pkg-config
configure:27663: result: /usr/bin/arm-linux-gnueabihf-pkg-config
configure:27731: checking pkg-config is at least version 0.16
configure:27734: result: yes
configure:27748: checking for GTK+ - version >= 3.0.0
configure:27875: result: no
configure:27903: arm-linux-gnueabihf-gcc -o conftest -D_GNU_SOURCE     conftest.c  -lm  >&5
conftest.c:53:10: fatal error: gtk/gtk.h: No such file or directory
   53 | #include <gtk/gtk.h>
      |          ^~~~~~~~~~~
compilation terminated.

```
Hope this is right :-|

> > There's a temp solution, or even a hack, i could use to be able to bypass the problem?
> You could try reverting the commit mentioned above -- does it work for you?

I really don't know how to do it :|
Does it means that i could overwrite the newer _configure.*_ with the older, working ones?

**Comment by vadz** [comment]
> these should be the relevant commands from _history_:

Ah, so you're just adding a foreign architecture. I didn't think of doing it like this, interesting, thanks. Is this the official way to cross-compile for ARM nowadays, rather than installing some sysroot image?

> Sorry. Let me try again:

Thanks, this is what I wanted to see, but now I'm lost: it looks like 3.1.4 configure doesn't use the correct flags (those returned by `--cflags` and shown above) neither, but it really ought to. I.e. it's clear that 3.1.7 fails because it doesn't use them, but how does 3.1.4 succeed?

> I really don't know how to do it :|

Normally it would be just `git revert -n 893ebbab0c5c569e68c0e8c8337e7b79cbc302eb`, but I've checked and this doesn't revert cleanly any more, at least in 3.2.1, due to other changes since then. The simplest is probably to find this part of configure:
```sh
if test "$build" != "$host"; then


                if test -z "$PKG_CONFIG_LIBDIR"; then
        PKG_CONFIG_LIBDIR=/dev/null
        export PKG_CONFIG_LIBDIR
    fi
fi
```
and simply remove it. Or replace `PKG_CONFIG_LIBDIR` with something else, e.g. run
```
$ sed -i'' -e 's/PKG_CONFIG_LIBDIR/DUMMY_PKG_CONFIG_LIBDIR/g' configure
```
to prevent it from having effect.

But, to be honest, I'm not sure if it's going to fix the problem any more, as I don't understand what's going on here. I'll try to test this myself but won't be able to do it immediately.

**Comment by parduz** [comment]
> Ah, so you're just adding a foreign architecture. I didn't think of doing it like this, interesting, thanks. Is this the official way to cross-compile for ARM nowadays, rather than installing some sysroot image?

Honestly, i don't know 🤣 
I'm just an old Windows programmer trying to develop for a Beaglebone Black. I feel lost in Linux, and i just follow guides until i get something, and then trying apply a bit of brain.
On the old Debian 9 PC, 5 years ago, that was the only way i've found to compile for ARM (and it was bugged, so i had to save the files installed in that way, revert to the x68 ones, and manually re-create the arm folders).
Now on Windows 10 the WSL is really helpful, as if i make mistakes i can wipe the Debian "machine" away and restart from scratch, and i'm taking the same basic steps as 5 years ago, seeing that some bug has been solved.
(If i could, i'd cross-compile from Windows directly, but i don't know how to _configure_ the widgets without Linux. )

> won't be able to do it immediately

Well, 6:30PM here, so i'll wait to hear for you tomorrow :)

**Comment by parduz** [comment]
I confirm that 

> The simplest is probably to find this part of configure:
> 
> ```shell
> if test "$build" != "$host"; then
> 
> 
>                 if test -z "$PKG_CONFIG_LIBDIR"; then
>         PKG_CONFIG_LIBDIR=/dev/null
>         export PKG_CONFIG_LIBDIR
>     fi
> fi
> ```
> 
> and simply remove it.

this works.
I commented the block out and i've been able to _configure_ and _make_ the widgets on my two PCs, both v3.1.7 and v3.2.1.

**Comment by vadz** [comment]
FWIW I've been working on cross-compiling my own program in the meanwhile and it looks like there are 2 (exclusive) ways of using pkg-config for cross-compiling:

1. Use native `pkg-config` with `PKG_CONFIG_DIR=''` and `PKG_CONFIG_LIBDIR=$target_root/usr/lib/$target_triplet/pkgconfig`, i.e. disable using the default directories and use the directories from the target sysroot instead.
2. Use `$target_tripet-pkg-config` directly, without setting any environment variables.

I am not sure if we can reliably distinguish between the two, so it's probably best to let the person (or automated process) running configure to set things up correctly and define `PKG_CONFIG_XXX` only if necessary, so I'm going to revert 893ebbab0c (Disable use of build system pkg-config files when cross-compiling, 2020-12-13) if/when #22915 CI builds pass.
